_blog/2023-09-13-building-web-scraper-in-rust.mdx

```diff
@@ -0,0 +1,381 @@
+
+In this guide to web scraping in Rust, we will write a Rust web scraper that will scrape Amazon for Raspberry Pi products and get their prices, then store them in a PostgresQL database for further processing.
+
+Stuck or want to see the final code? The Github repository for this article can be found [here.](https://github.com/joshua-mo-143/reqshuttle)
+
+## Getting Started
+Let's make a new project by using `cargo shuttle init`. For this project we'll simply call it `webscraper` - you'll want the `Custom` web service, which will spawn a new Cargo project with `shuttle-runtime` added.
+
+Let's install our dependencies with the following one liner:
+```bash
+cargo add chrono reqwest scraper tracing shuttle-shared-db sqlx --features shuttle-shared-db/postgres,sqlx/runtime-tokio-native-tls,sqlx/postgres
+```
+We'll also want to install `sqlx-cli`, which is a useful tool for managing our SQL migrations. We can install it by running te following:
+```bash
+cargo install sqlx-cli
+```
+If we then use `sqlx migrate add schema` in our project folder, we'll then get our SQL migration file, which can be found in the `migrations` folder! The file will be formatted with the date and time at which the migration was created, and then the name we gave it (in this case, `schema`). For our purposes, here are the migrations we'll be using:
+```sql
+CREATE TABLE IF NOT EXISTS products (
+        id SERIAL PRIMARY KEY,
+        name VARCHAR NOT NULL,
+        price VARCHAR NOT NULL,
+        old_price VARCHAR,
+        link VARCHAR,
+        scraped_at DATE NOT NULL DEFAULT CURRENT_DATE   
+);
+```
+
+Before we get started, we'll want to make a struct that implements `shuttle_runtime::Service`, which is an async trait. We'll also want to set our user agent so that there is less chance of us getting blocked. Thankfully, we can do all of this by returning a struct in our main function, like so:
+```rust
+use reqwest::Client;
+use tracing::error;
+use sqlx::PgPool;
+
+pub struct CustomService {
+    ctx: Client,
+    db: PgPool
+}
+
+// Set up our user agent
+const USER_AGENT: &str = "Mozilla/5.0 (iPad; CPU OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148";
+
+// note that we add our Database as an annotation here so we can easily get it provisioned to us
+#[shuttle_runtime::main]
+async fn main(
+  #[shuttle_shared_db::Postgres] db: PgPool
+) -> Result<CustomService, shuttle_runtime::Error> {
+// automatically attempt to do migrations
+// we only create the table if it doesn't exist which prevents data wiping
+ sqlx::migrate!().run(&db).await.expect("Migrations failed");
+// initialise Reqwest client here so we can add it in later on
+    let ctx = Client::builder().user_agent(USER_AGENT).build().unwrap();
+    Ok(CustomService { ctx, db })
+}
+
+#[shuttle_runtime::async_trait]
+impl shuttle_runtime::Service for CustomService {
+    async fn bind(mut self, _addr: SocketAddr) -> Result<(), shuttle_runtime::Error> {
```

Comment:
amended in next commit

I've elected to just add the full type to the function instead of importing since the program currently does not use a HTTP bound service